### DISCUSSION TITLE
Serverless Function Execution on Every Route Change with SSG and ISR

### DISCUSSION THREAD

**Discussion by jadamconnor** [vercel/next.js#33568] [Help] [answered]
We have a Next.js app that requires full page reloads on every route change in order to load ads, so we are not using the next/link API. In addition to this, we're using getStaticPaths and getStaticProps with blocking: true and revalidation in order to implement SSG and ISR. We're hosting our app on Vercel, and I'm noticing some concerning behavior. We end up with billable serverless function execution on every page, even on pages that should be statically generated. Is this a result of not using next/link and forcing page reloads on each route change?

**Comment by leerob** [comment, marked as answer]
It sounds like your `revalidate` time is low, which would be causing those function calls to be invocated in the background. You might consider raising that revalidation time if you're concerned about function execution time – or taking a look at how long those functions are running. For example, if they're running for 5 seconds, you might consider optimizing / caching the code in that function.

**Reply by jadamconnor** [reply]
Yes. Revalidate was set to 1 second. That was the result of me misunderstanding how revalidate worked. I was under the impression that getStaticProps would only be invocated if a check of some kind determined that the page data was stale. (I admit I have no idea what that check would be now that I think of it.) But it seems that revalidate simply means that the page will execute getStaticProps at the interval specified by revalidate. Is this correct?

Anyway, I increased revalidate to 3600, and I noticed immediately that the pages were generally only executing functions after the first execution for an hour or so. But after the second validation they began revalidating at unexpected intervals. I've attached an image for reference. You can see in the image multiple static requests to the slug. But then you see lambdas at interval < our current revalidation value (3600).

![image](https://user-images.githubusercontent.com/23176631/150669340-ad372c4a-e665-4327-bc2f-2e21e00bd9f5.jpeg)

This causes me to wonder whether there is some special case in which a serverless function will execute outside of revalidation or build. Any insight here would be helpful!

**Reply by jadamconnor** [reply]
One other thing I noticed is that after increasing the revalidate value, build speed has increased significantly (~3-5x) and there are no more build failures--we were getting a lot of those. The majority of the build fails had to do with the API returning an empty JSON object for a given page. Might this have something to do with attempted revalidations during build?
